File: step17example/src/main/java/com/example/step17example/MainActivity.java
package com.example.step17example;

import static com.example.step17example.AppConstant.BASE_URL;
import static com.example.step17example.AppConstant.REQUEST_TODO_ADDDATA;
import static com.example.step17example.AppConstant.REQUEST_TODO_DELETE;
import static com.example.step17example.AppConstant.REQUEST_TODO_INSERT;
import static com.example.step17example.AppConstant.REQUEST_TODO_LIST;
import static com.example.step17example.AppConstant.REQUEST_TODO_UPDATE;

import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.example.step17example.databinding.ActivityMainBinding;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
/*
    view bindiing 사용하는 방법

    1. build.gradle 파일에 아래의 설정 추가

        buildFeatures{
            viewBinding = true
        }
    2. 우상단에 sync now 링크를 눌러서 설정이 적용되도록 한다.
    3. layout.xml 문서의 이름대로 클래스가 만들어진다.
       예를 들어 activity_main.xml 문서면 ActivityMainBinding클래스
                activity_sub.xml 문서면 ActivitySubBinding 클래스
 */

public class MainActivity extends AppCompatActivity implements Util.RequestListener, AdapterView.OnItemLongClickListener, AbsListView.OnScrollListener {

    ActivityMainBinding binding;
    ArrayList<Todo> list;
    TodoAdapter adapter;
    Date date;
    String msg;
    String num;
    Boolean isNext = true;
    private Boolean lastList = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        // setContentView(R.layout.activity_main);

        // R.Layout.activity_main.xml 문서를 전개해서 View를 만들기
       binding = ActivityMainBinding.inflate(getLayoutInflater());
        // 전개된 Layout에서 root를 얻어내서 화면 구성을 한다.( 여기서는 LinearLayout이다. )
        setContentView(binding.getRoot());

        list = new ArrayList<>();


        binding.addBtn.setOnClickListener((view)->{
            date = new Date();
            msg = binding.inputText.getText().toString();
            Map<String,String> map1 = new HashMap<>();
            map1.put("content", msg);
            Util.sendPostRequest(REQUEST_TODO_INSERT, BASE_URL+"todo/insert", map1, this);
            binding.inputText.setText("");
            Util.hideKeyboard(this);
        });

        // ListView를 오랬동안 클릭했을 때
        binding.listView.setOnItemLongClickListener(this);
        binding.listView.setOnScrollListener(this);
        binding.progressbar.setVisibility(View.GONE);
    }

    @Override
    protected void onStart() {
        super.onStart();

        Map<String,String> map = new HashMap<>();
        num = "1";
        map.put("num", num);

        Util.sendPostRequest(REQUEST_TODO_LIST, BASE_URL+"todo/list", map, this);

    }

    @Override
    public void onSuccess(int requestId, Map<String, Object> result)  {
        if(requestId == REQUEST_TODO_LIST){
            String data = (String)result.get("data");
            try {
                JSONObject array = new JSONObject(data);
                JSONArray array2 = array.getJSONArray("result");
                isNext = array.getBoolean("isNext");

                for(int i=0; i<array2.length(); i++){
                    Todo todo = new Todo();
                    todo.setNum(array2.getJSONObject(i).getInt("num"));
                    todo.setContent(array2.getJSONObject(i).getString("content"));
                    todo.setRegdate(array2.getJSONObject(i).getString("regdate"));
                    list.add(todo);
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
            adapter = new TodoAdapter(this, R.layout.listview_cell, list);
            binding.listView.setAdapter(adapter);
        }else if(requestId == REQUEST_TODO_INSERT){
            Util.sendPostRequest(REQUEST_TODO_ADDDATA, BASE_URL+"todo/getAddData", null, this);
        }else if(requestId == REQUEST_TODO_ADDDATA){
            String data = (String)result.get("data");
            try {
                JSONObject json = new JSONObject(data);
                Todo todo = new Todo();
                todo.setNum(json.getInt("num"));
                todo.setContent(json.getString("content"));
                todo.setRegdate(json.getString("regdate"));
                list.add(0, todo);
                adapter.notifyDataSetChanged();
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }else if(requestId == REQUEST_TODO_UPDATE){
            binding.progressbar.setVisibility(View.GONE);

            String data = (String)result.get("data");
            JSONObject array;
            try {
                array = new JSONObject(data);
                JSONArray array2 = array.getJSONArray("result");
                isNext = array.getBoolean("isNext");

                for(int i=0; i<array2.length(); i++){
                    Todo todo = new Todo();
                    todo.setNum(array2.getJSONObject(i).getInt("num"));
                    todo.setContent(array2.getJSONObject(i).getString("content"));
                    todo.setRegdate(array2.getJSONObject(i).getString("regdate"));
                    list.add(todo);
                    adapter.notifyDataSetChanged();
                }

            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onFail(int requestId, Map<String, Object> result) {
        Log.d("#### json 문자열 ####", "fail");

    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {

        new AlertDialog.Builder(this)
                .setTitle("알림")
                .setMessage("삭제하시겠습니까?")
                .setPositiveButton("네", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                         /*
                            view는 클릭한 cell의 View
                            position은 클릭한 cell의 인덱스
                            id는 클릭한 cell 모델의 primary key 값
                         */
                        // 삭제할 할일의 primary key 값을 Map에 담고
                       HashMap<String, String> map = new HashMap<>();
                        map.put("num", Integer.toString(list.get(position).getNum()));
                        Util.sendPostRequest(REQUEST_TODO_DELETE, BASE_URL+"todo/delete", map, MainActivity.this);

                        list.remove(position);
                        adapter.notifyDataSetChanged();
                    }
                })
                .setNegativeButton("아니오", null)
                .create()
                .show();

        return false;
    }

    @Override
    public void onScrollStateChanged(AbsListView view, int scrollState) {
        if(isNext == true && scrollState == AbsListView.OnScrollListener.SCROLL_STATE_IDLE && lastList ){
            binding.progressbar.setVisibility(View.VISIBLE);

            Map<String,String> map = new HashMap<>();
            int next = Integer.parseInt(num)+1;
            num = Integer.toString(next);

            map.put("num", num);

            Util.sendPostRequest(REQUEST_TODO_UPDATE, BASE_URL+"todo/list", map, this);
        }

    }

    @Override
    public void onScroll(AbsListView view, int firstVisibleItem, int visibleItemCount, int totalItemCount) {
        lastList = (totalItemCount > 0 ) && (firstVisibleItem + visibleItemCount >= totalItemCount);
    }
}

File: step13sharedpref2/src/main/java/com/example/step13sharedpref2/MainActivity.kt
package com.example.step13sharedpref2

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import androidx.preference.PreferenceManager

/*
    App에서 문자열을 영구 저장하는 방법 ( 영구 저장이란 앱을 종료하고 다시 시작해도 불러올 수 있는 문자열)

    1. 파일 입출력을 이용해서 저장
    2. android 내장 data base를 이용해서 저장 => SQLite DataBase
    3. SharedPreference를 이용해서 저장(느리지만 간단히 저장하고 불러올 수 있다)
       내부적으로 xml 문서를 만들어서 문자열을 저장하고 불러온다.
       저장된 문자열을 boolean, int, double, String type으로 변환해서 불러올 수 있다.

 */

class MainActivity : AppCompatActivity(), View.OnClickListener {
    /*
         java에서는 field를 선언만 하면 자동으로 null로 초기화 됨.
         kotlin에서는 null이 가능한 field를 만들어서 명시적으로 넣어줘야한다.
     */
    var editText: EditText? = null

    // onCreate()메소드 재정의
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // res/layout/activity_main.xml 문서를 전개해서 화면 구성하기
        setContentView(R.layout.activity_main)

        // EditText 객체의 참조값 얻어오기
        editText = findViewById<EditText>(R.id.editText)
        // Button 객체의 참조값 얻어오기
        // val saveBtn = findViewById<Button>(R.id.saveBtn)
        val saveBtn: Button = findViewById(R.id.saveBtn)
        saveBtn.setOnClickListener(this)

        val readBtn = findViewById<Button>(R.id.readBtn)
        /*
        readBtn.setOnClickListener(object:View.OnClickListener{
            override fun onClick(p0: View?) {
            }
        })
         */
        // 위의 코드를 간략히 표현하면 아래와 같다.
        readBtn.setOnClickListener{
            val pref:SharedPreferences = getSharedPreferences("info", Context.MODE_PRIVATE)
            // "msg"라는 키값으로 저장된 문자열 읽어오기, 없다면 default값이 읽어와 진다.
            val msg = pref.getString("msg", "")
            /*
                여기서 this는 MainActivity객체를 가리킨다.
                java에서는 익명클래스 안에서 바깥클래스 객체의 참조값을 가리키려면
                MainActivity.this와 같이 클래스명을 명시했어야한다.

                kotlin에서도 익명 클래스 안에서 바깥 클래스 객체의 참조값을 가리키려면
                this@MainClass와 같이 클래스명을 명시하면된다.
             */
            Toast.makeText(this, msg,Toast.LENGTH_SHORT).show()
        }
    }

    override fun onStart() {
        super.onStart()

        val pref = PreferenceManager.getDefaultSharedPreferences(this)
        // 액티비티가 활성화되는 시점에 설정에 저장된 값을 읽어오고 싶다면 여기서 작업하면 된다.
        val signature = pref.getString("signature","")
        val reply = pref.getString("reply", "")
        val sync = pref.getBoolean("sync", false)

        Toast.makeText(this, "signature:${signature}, reply:${reply}, sync:${sync}", Toast.LENGTH_LONG).show()
    }

    // 저장 버튼을 누르면 호출되는 메소드
    override fun onClick(p0: View?) {
        //EditText에 입력한 문자열 읽어오기
        val msg = editText?.text.toString() // null이 가능한 변수나 필드의 값을 참조할 때는 ?가 필요하다.
        // SharedPreferences의 참조값 얻어오기
        val pref: SharedPreferences = getSharedPreferences("info", Context.MODE_PRIVATE)
        // 에디터 객체의 참조값 얻어오기
        val editor: SharedPreferences.Editor = pref.edit()
        // 에디터 객체를 이용해서 문자열을 key:value 형태로 영구 저장할 수 있다.
        editor.putString("msg", msg)
        editor.commit()

        AlertDialog.Builder(this)
            .setMessage("저장했습니다.")
            .setNeutralButton("확인",null)
            .create()
            .show()
    }
    // 옵션 메뉴를 구성하는 메소드
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        // menu/menu_main.xml 문서를 전개해서 메뉴 구성하기
        // in java => getMenuInflater().inflate()
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    // 옵션 아이템을 선택했을 때 호출되는 메소드
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // 선택한 메뉴의 아이디 읽어오기
        val id = item.itemId
        // 만일 설정 메뉴를 선택했다면
        if(id == R.id.setting){
            // kotlin에서 특정 클래스 type은 클래스명::class.java라고 해야한다.
            val intent = Intent(this, SettingsActivity::class.java)
            // SettingsActivity를 시작하겠다는 의도를 가지고 있는 Intent 객체를 이용해서
            // 액티비티 시작 시키기
            startActivity(intent)
        }
        return super.onOptionsItemSelected(item)
    }
}